### PR TITLE
Add EIP-5341: EIP NFT referral proposal

### PR DESCRIPTION
**Pull request by cdpiano**
When opening a pull request to submit a new EIP, please use the suggested template: https://github.com/ethereum/EIPs/blob/master/eip-template.md

We have a GitHub bot that automatically merges some PRs. It will merge yours immediately if certain criteria are met:

 - The PR edits only existing draft PRs.
 - The build passes.
 - Your GitHub username or email address is listed in the 'author' header of all affected PRs, inside <triangular brackets>.
 - If matching on email address, the email address is the one publicly listed on your GitHub profile.

### CODE REVIEW

File: EIPS/eip-NFT-Referral-Program.md [hunk 1]
```diff
@@ -0,0 +1,117 @@
+---
+eip: <to be assigned>
```
Comment (MicahZoltu) [inline EIPS/eip-NFT-Referral-Program.md:2]: Please rename the file to `EIPS/eip-5341.md` as well.
```suggestion
eip: 5341
```

**Review by MicahZoltu — COMMENTED** [review]
Please see the Files tab of GitHub and address all of the comments by the bot.

File: EIPS/eip-5341.md [hunk 2]
```diff
@@ -0,0 +1,117 @@
+---
+eip: 5341
+title: NFT Referral Program
+description: An interface for defining NFT referral program from which referrers share referral code and earn rebates.
```
Comment (Pandapip1) [inline EIPS/eip-5341.md:4]:
```suggestion
description: An interface for defining a program where referrers share referral codes and earn rebates
```

File: EIPS/eip-5341.md [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+discussions-to: https://ethereum-magicians.org/t/new-proposal-nft-referral-program/10121
+status: Draft
+type: Standards Track
+category: ERC
+created: 2022-07-25
+requires: 721
+---
+
+## Abstract
+An standard interface for defining NFT referral program from which referrers share referral code and earn a percentage of mint cost from referees who mint by using the referral code.
```
Comment (Pandapip1) [inline EIPS/eip-5341.md:15]:
```suggestion
This EIP describes an interface for defining a program where referrers share referral codes and earn rebates.
```
Needs to explain solution a bit more, too.

File: EIPS/eip-5341.md [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+
+## Motivation
+In real world we all need to rent or buy a house. The goal of renting or buying a house is not always to trade or invest, but to utilize it as a living space as well. Existing residents get bonus after they invite a new tenant to join, before which they spent time introducing the apartment to potential new tenants and answering their questions from different angles. These existing residents significantly contribute to the success of an apartment rental, building a large and interactive community.
```
Comment (Pandapip1) [inline EIPS/eip-5341.md:18]:
```suggestion
In the real world, we all need to rent or buy a house. The goal of renting or buying a house is not always to trade or invest, but to utilize it as a living space as well. Existing residents get a bonus after they invite a new tenant to join, before which they spent time introducing the apartment to potential new tenants and answering their questions from different angles. These existing residents significantly contribute to the success of an apartment rental, building a large and interactive community.
```

File: EIPS/eip-5341.md [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+
+It's likewise for NFT space. Not only team members of a project, it's the influencers and the strong believers of a project who introduce the project to more people and help new people onboard. They are like ambassadors of the project and contribute a lot toe the project, but currently there isn't a good way to incentivize them besides their true passion for the project and community. We would like to propose a standard interface for NFT projects to give referral bonus to these ambassadors.
```
Comment (Pandapip1) [inline EIPS/eip-5341.md:20]:
```suggestion
This happens in the NFT space too. Influencers and devoted users often encourage other people to participate in a project, despite a lack of incentive beyond passion. Referral programs provide this needed incentive.
```

File: EIPS/eip-5341.md [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+
+Enabling more NFT projects to unify on a referral program standard will benefit the entire NFT ecosystem and we can see this makes a larger impact on non-transferrable NFTs. Currently NFT is still considered something money-oriented and some people who mint are just to flip and take profit. However for non-transferrable NFTs the utility attribute will be enhanced and the financial attribute weakened, just like apartment rental where you don't usually flip the apartment but live inside it. Referral program may contribute to the success more in the space of non-transferrable NFTs.
```
Comment (Pandapip1) [inline EIPS/eip-5341.md:22]:
```suggestion
```
This is a repetition of things that have previously been mentioned.

File: EIPS/eip-5341.md [hunk 7]
```diff
@@ -0,0 +1,117 @@
...
+## Specification
+The key words “MUST”, “MUST NOT”, “REQUIRED”, “SHALL”, “SHALL NOT”, “SHOULD”, “SHOULD NOT”, “RECOMMENDED”, “MAY”, and “OPTIONAL” in this document are to be interpreted as described in RFC 2119.
+
+### Contract Interface
+
+```solidity
+interface INFTReferralProgram {
+    /// @notice Get the referrer address of a tokenId
+    /// @param tokenId The tokend id of the NFT
+    function referrerOf(uint256 tokenId) external view returns (address);
```
Comment (Pandapip1) [inline EIPS/eip-5341.md:33]: How is this determined?

File: EIPS/eip-5341.md [hunk 8]
```diff
@@ -0,0 +1,117 @@
...
+Owner can optionally set the referral bonus for referrer once a referee mints. We don't enforce a standard for referral code as there are multiple ways to do it and the easiest way is to use the referrer's address as referral code. 
+
+### Record the referrer of each NFT
+This EIP records the referrer of each NFT, which helps developers write payment sent logic as well as this data can be used for numerious scenarios after the sales. Referral information can be considered as part of social elements that can be utilized by NFT projects in the long term. 
+
+### Simple payment to a single address
+This EIP does not specify the method of sending payment to referrers as there can be various ways of doing it. Attempting to do this as part of this standard would increase the implementation complexity and increase gas costs. This ERC should be considered a minimal, gas-efficient building block for further innovation in referral program.
+
+## Backwards Compatibility
+This standard is fully ERC-721 compatible by adding an extension function set.
```
Comment (Pandapip1) [inline EIPS/eip-5341.md:54]:
```suggestion
This standard is backwards compatible with [EIP-721](./eip-721.md)
```

File: EIPS/eip-5341.md [hunk 9]
```diff
@@ -0,0 +1,117 @@
...
+
+Besides, new functions introduced in this standard have many similarities with the existing functions in ERC-721, which helps developers to easily adopt the standard quickly.
```
Comment (Pandapip1) [inline EIPS/eip-5341.md:56]:
```suggestion
```
If anything, this should be put in the Rationale.

File: EIPS/eip-5341.md [hunk 10]
```diff
@@ -0,0 +1,117 @@
...
+        returns (uint256)
+    {
+        return (salesPrice * _referralFraction) / _referralFeeDenominator();
+    }
+
+    /// @notice The denominator with which to interpret the fee set in {_setReferralFraction} as a fraction of the sale price.
+    /// Defaults to 10000 so fees are expressed in basis points, but may be customized by an override.
+    function _referralFeeDenominator() internal pure virtual returns (uint96) {
+        return 10000;
+    }
```
Comment (Pandapip1) [inline EIPS/eip-5341.md:110]:
```suggestion
    }
}
```

File: EIPS/eip-5341.md [hunk 11]
```diff
@@ -0,0 +1,117 @@
...
+```
+
+## Security Considerations
+The EIP sets the referral fraction and the token to referrer mapping. The owner can use them to transfer referral bonus to referrer when referee mints.
```
Comment (Pandapip1) [inline EIPS/eip-5341.md:114]: So, you're saying that the project maintainers must be trusted to pay referrers out? If so, could you make that more explicit?

File: EIPS/eip-5341.md [hunk 12]
```diff
@@ -0,0 +1,117 @@
...
+    // The percentage of sale price referrer obtains
+    uint96 private _referralFraction;
+
+    // Mapping from tokenId to the referrer address
+    mapping(uint256 => address) internal _referrers;
+
+    /// @notice Set the referrer of a NFT
+    /// @param tokenId The tokend id of the NFT
+    /// @param referrer The referrer address of the token
+    function _setReferrer(uint256 tokenId, address referrer) internal virtual {
```
Comment (Pandapip1) [inline EIPS/eip-5341.md:70]: How are the dues tracked?

**Review by Pandapip1 — CHANGES_REQUESTED** [review]

File: EIPS/eip-5341.md (hunk 7 above)
Comment (xinbenlv) [reply EIPS/eip-5341.md:32]: +1 
Probably this EIP shall also specify a `refer` function to set the information of the referrals to be queried by `referrerOf` and `referralFee`

**Pull request closed without merging**